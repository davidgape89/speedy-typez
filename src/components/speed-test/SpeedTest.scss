* {
    margin-bottom: 2rem;
}

.speed-test {
    display: flex;
    flex-direction: column;
}

.speed-test__text {
  font-size: 6rem;
  height: 19rem;
  overflow-y: hidden;

  span {
    border-radius: 1rem;
    padding: 2px;
  }

  span.is-active {
    background-color: lighten(grey, 30);
  }

  span.is-invalid {
    background-color: lighten(red, 30);
  }

  span.is-valid {
    background-color: lighten(green, 50);
  }
}
