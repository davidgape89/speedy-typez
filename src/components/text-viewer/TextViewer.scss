$textSize: 2rem;

.text-viewer {
  font-size: $textSize;
  height: $textSize * 3.1;
  overflow: hidden;

  span {
    border-radius: 1rem;
    padding: 2px;
  }

  span.is-active {
    background-color: lighten(grey, 30);
  }

  span.is-invalid {
    background-color: lighten(red, 30);
  }

  span.is-valid {
    background-color: lighten(green, 50);
  }
}
