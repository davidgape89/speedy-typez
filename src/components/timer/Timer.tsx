import React, { useState, useEffect, forwardRef, useImperativeHandle } from 'react';

export enum TimerDirection {
  ASC = 1,
  DESC = -1,
}

interface TimerProps {
  initTime: number;
  onTimeOver: () => void;
  isRunning?: boolean;
  direction?: TimerDirection;
}

export interface TimerRef {
  reset: () => void;
}

export const Timer = forwardRef(({
    isRunning = true,
    initTime,
    direction = TimerDirection.DESC,
    onTimeOver
  }: TimerProps, ref) => {
  const [time, setTime] = useState(initTime);
  const [timeoutRef, setTimeoutRef] = useState();
  const minutes = Math.floor(time / 60).toString();
  const seconds = (time % 60).toString();

  if ( isRunning && !timeoutRef) {
    if (time === 0) {
      onTimeOver();
    } else {
      setTimeoutRef(
        setTimeout(() => {
          setTime(time => time + direction)
          setTimeoutRef(null);
        }, 1000)
      );
    }
  }

  useImperativeHandle(ref, () => ({
    reset() {
      setTime(initTime);
    }
  }));

  useEffect(() => {
    if (!isRunning && timeoutRef) {
      clearTimeout(timeoutRef);
      setTimeoutRef(null);
    }

    return () => clearTimeout(timeoutRef);
  }, [isRunning, timeoutRef]);

  return (
    <span>
      {minutes.length === 1 ? `0${minutes}`: minutes}:
      {seconds.length === 1 ? `0${seconds}`: seconds}
    </span>
  )
});
