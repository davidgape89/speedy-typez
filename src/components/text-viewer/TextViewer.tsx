import React, { useEffect } from 'react';
import './TextViewer.scss';

export interface TextViewerProps {
  words: string[];
  inputStack: string[];
  activeInput: string;
}

export function TextViewer(props: TextViewerProps) {
  const {
    words,
    inputStack,
    activeInput
  } = props;

  useEffect(() => {
    const activeEl: HTMLElement | null = document.querySelector('.is-active');
    if(activeEl) activeEl.scrollIntoView();
  }, [inputStack]);

  function getWordStatus(i: number) {
    if (inputStack[i]) {
      if(words[i] === inputStack[i]) return "is-valid";
      else return "is-invalid";
    } else if (inputStack.length === i) {
      if(words[i].startsWith(activeInput)) return "is-active";
      else return "is-invalid";
    }
  }

  return (
    <div className="text-viewer">
      {words.map((word, i) => (
        <React.Fragment key={i}>
          <span
            className={getWordStatus(i)}>
            {word}
          </span>
          <span>{' '}</span>
        </React.Fragment>
      ))}
    </div>
  );
}
