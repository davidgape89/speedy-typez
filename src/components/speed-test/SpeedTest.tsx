import React, { useEffect, useRef, useState } from 'react';
import { Timer, TimerRef } from '../timer/Timer';
import { englishText } from './words';
import { TextViewer } from '../text-viewer/TextViewer';

import './SpeedTest.scss';

export default function SpeedTest() {
  const timerRef = useRef<TimerRef>();
  const [input, setInput] = useState<string>('');
  const [isRunning, setIsRunning] = useState<boolean>(false);
  const [isDisabled, setIsDisabled] = useState<boolean>(false);
  const [inputStack, setInputStack] = useState<string[]>([]);
  const words = englishText.replace(/\n/g, '').split(' ');

  useEffect(() => {
    if (input.length > 0 && !isRunning) {
      setIsRunning(true);
    }
  }, [input, isRunning]);

  function reset() {
    setInput('');
    setInputStack([]);
    setIsDisabled(false);
    setIsRunning(false);
    if (timerRef && timerRef.current) timerRef.current.reset();
  }

  function handleTimeOver() {
    setIsDisabled(true);
  }

  function handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>) {
    switch (e.keyCode) {
      // Space hit
      case 32:
        e.preventDefault();
        setInputStack([...inputStack, input]);
        setInput('');
        break;
      // Backspace
      case 8:
        if (inputStack.length > 0 && !input.length) {
          e.preventDefault();
          const lastIndex = inputStack.length - 1;
          setInput(inputStack[lastIndex]);
          setInputStack(inputStack.slice(0, lastIndex))
        }
        break;
      // Escape restarts
      case 27:
        reset();
        break;
    }
  }

  return (
    <div className="speed-test">
      <TextViewer
        words={words}
        inputStack={inputStack}
        activeInput={input}
      />
      <div className="speed-test__input-bar">
        <input
          data-testid="input-field"
          type="text"
          disabled={isDisabled}
          value={input}
          onChange={e => setInput(e.target.value)}
          onKeyDown={handleKeyPress}
        />
        <Timer
          data-testid="timer"
          ref={timerRef}
          initTime={60 * 1}
          isRunning={isRunning}
          onTimeOver={handleTimeOver}
        />
      </div>
      {isDisabled && <div>
        Your word count is {inputStack.length / 2}!
      </div>}
    </div>
  )
}
