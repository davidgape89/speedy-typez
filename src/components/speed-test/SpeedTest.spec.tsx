import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SpeedTest from './SpeedTest';
import { TextViewer } from '../text-viewer/TextViewer';
import { Timer } from '../timer/Timer';

jest.mock('../text-viewer/TextViewer', () => ({
  TextViewer: jest.fn(() => <div>Text viewer</div>),
}));
jest.mock('../timer/Timer', () => ({
  Timer: jest.fn(() => <div>Timer</div>),
}));

test('renders correctly', () => {
  const wrapper = render(<SpeedTest />);

  expect(wrapper.container.firstChild).toMatchSnapshot();
});

test('starts running once something is typed', () => {
  const wrapper = render(<SpeedTest />);
  const input = wrapper.getByTestId('input-field');

  fireEvent.change(input, {
    target: {
      value: 'something',
    },
  });

  jest.runAllTimers();

  expect(Timer).toHaveBeenCalledWith({isRunning: true});
  // back to enzyme?
  //wrapper.getByTestId('timer').getAttribute('isRunning').toBe(true);
});
