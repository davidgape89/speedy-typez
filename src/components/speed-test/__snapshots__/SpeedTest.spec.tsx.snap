// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="speed-test"
>
  <div
    class="text-viewer"
  >
    <span
      class="is-active"
    >
      hey
    </span>
    <span>
       
    </span>
    <span>
      how
    </span>
    <span>
       
    </span>
    <span>
      are
    </span>
    <span>
       
    </span>
    <span>
      you
    </span>
    <span>
       
    </span>
    <span>
      this
    </span>
    <span>
       
    </span>
    <span>
      is
    </span>
    <span>
       
    </span>
    <span>
      a
    </span>
    <span>
       
    </span>
    <span>
      list
    </span>
    <span>
       
    </span>
    <span>
      of
    </span>
    <span>
       
    </span>
    <span>
      different
    </span>
    <span>
       
    </span>
    <span>
      english
    </span>
    <span>
       
    </span>
    <span>
      words
    </span>
    <span>
       
    </span>
    <span>
      just
    </span>
    <span>
       
    </span>
    <span>
      to
    </span>
    <span>
       
    </span>
    <span>
      perform
    </span>
    <span>
       
    </span>
    <span>
      a
    </span>
    <span>
       
    </span>
    <span>
      typing
    </span>
    <span>
       
    </span>
    <span>
      speed
    </span>
    <span>
       
    </span>
    <span>
      test
    </span>
    <span>
       
    </span>
    <span>
      no
    </span>
    <span>
       
    </span>
    <span>
      commas
    </span>
    <span>
       
    </span>
    <span>
      no
    </span>
    <span>
       
    </span>
    <span>
      special
    </span>
    <span>
       
    </span>
    <span>
      signs
    </span>
    <span>
       
    </span>
    <span>
      or
    </span>
    <span>
       
    </span>
    <span>
      anything
    </span>
    <span>
       
    </span>
    <span>
      like
    </span>
    <span>
       
    </span>
    <span>
      that
    </span>
    <span>
       
    </span>
    <span>
      just
    </span>
    <span>
       
    </span>
    <span>
      words
    </span>
    <span>
       
    </span>
    <span>
      like
    </span>
    <span>
       
    </span>
    <span>
      for
    </span>
    <span>
       
    </span>
    <span>
      instance
    </span>
    <span>
       
    </span>
    <span>
      keyboard
    </span>
    <span>
       
    </span>
    <span>
      mouse
    </span>
    <span>
       
    </span>
    <span>
      laptop
    </span>
    <span>
       
    </span>
    <span>
      lamp
    </span>
    <span>
       
    </span>
    <span>
      chair
    </span>
    <span>
       
    </span>
    <span>
      microphone
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      guess
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      am
    </span>
    <span>
       
    </span>
    <span>
      doing
    </span>
    <span>
       
    </span>
    <span>
      myself
    </span>
    <span>
       
    </span>
    <span>
      no
    </span>
    <span>
       
    </span>
    <span>
      favour
    </span>
    <span>
       
    </span>
    <span>
      by
    </span>
    <span>
       
    </span>
    <span>
      including
    </span>
    <span>
       
    </span>
    <span>
      long
    </span>
    <span>
       
    </span>
    <span>
      words
    </span>
    <span>
       
    </span>
    <span>
      since
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      longer
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      words
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      fewer
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      will
    </span>
    <span>
       
    </span>
    <span>
      be
    </span>
    <span>
       
    </span>
    <span>
      able
    </span>
    <span>
       
    </span>
    <span>
      to
    </span>
    <span>
       
    </span>
    <span>
      write
    </span>
    <span>
       
    </span>
    <span>
      in
    </span>
    <span>
       
    </span>
    <span>
      this
    </span>
    <span>
       
    </span>
    <span>
      text
    </span>
    <span>
       
    </span>
    <span>
      this
    </span>
    <span>
       
    </span>
    <span>
      is
    </span>
    <span>
       
    </span>
    <span>
      quite
    </span>
    <span>
       
    </span>
    <span>
      complicated
    </span>
    <span>
       
    </span>
    <span>
      what
    </span>
    <span>
       
    </span>
    <span>
      am
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      doing
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      just
    </span>
    <span>
       
    </span>
    <span>
      remembered
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      need
    </span>
    <span>
       
    </span>
    <span>
      to
    </span>
    <span>
       
    </span>
    <span>
      implement
    </span>
    <span>
       
    </span>
    <span>
      a
    </span>
    <span>
       
    </span>
    <span>
      line
    </span>
    <span>
       
    </span>
    <span>
      shifter
    </span>
    <span>
       
    </span>
    <span>
      for
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      display
    </span>
    <span>
       
    </span>
    <span>
      of
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      text
    </span>
    <span>
       
    </span>
    <span>
      in
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      speed
    </span>
    <span>
       
    </span>
    <span>
      test
    </span>
    <span>
       
    </span>
    <span>
      component
    </span>
    <span>
       
    </span>
    <span>
      but
    </span>
    <span>
       
    </span>
    <span>
      I
    </span>
    <span>
       
    </span>
    <span>
      guess
    </span>
    <span>
       
    </span>
    <span>
      that
    </span>
    <span>
       
    </span>
    <span>
      is
    </span>
    <span>
       
    </span>
    <span>
      not
    </span>
    <span>
       
    </span>
    <span>
      urgent
    </span>
    <span>
       
    </span>
    <span>
      since
    </span>
    <span>
       
    </span>
    <span>
      the
    </span>
    <span>
       
    </span>
    <span>
      test
    </span>
    <span>
       
    </span>
    <span>
      is
    </span>
    <span>
       
    </span>
    <span>
      perfectly
    </span>
    <span>
       
    </span>
    <span>
      doable
    </span>
    <span>
       
    </span>
    <span>
      without
    </span>
    <span>
       
    </span>
    <span>
      it
    </span>
    <span>
       
    </span>
  </div>
  <div
    class="speed-test__input-bar"
  >
    <input
      data-testid="input-field"
      type="text"
      value=""
    />
    <span
      class="timer"
    >
      01
      :
      00
    </span>
  </div>
</div>
`;
